src/Swarm/Game/Step.hs

```diff
@@ -1065,6 +1065,92 @@ seedProgram minTime randTime thing =
     selfdestruct
   |]
 
+-- | A system program for a "combustion robot", to burn an entity
+--   after it is ignited.
+--
+-- Creates sub-partitions (of 10-tick duration) of the "stages"
+-- of combustion for opportunities to light adjacent entities on fire.
+combustionProgram :: Integer -> Integer -> Text -> Double -> ProcessedTerm
+combustionProgram minTime randTime thing spreadProbabilityPerTick =
+  [tmQ|
+    def max = \a. \b.
+      if (a > b) {a} {b};
+      end;
+
+    def trySpread = \denom. \d.
+      try {
+        r <- random denom;
+        if (r == 0) {
+          ignite d;
+          noop;
+        } {};
+      } {};
+      end;
+
+    def tryDirs = \denom.
+      trySpread denom forward;
+      trySpread denom left;
+      trySpread denom back;
+      trySpread denom right;
+      end;
+
+
+    try {
+      r <- random (1 + $int:randTime);
+      wait (r + $int:minTime);
+      appear "X";
+      r <- random (1 + $int:randTime);
+      let spreadableDuration = r + $int:minTime in
+      let segmentCount = max 1 $ spreadableDuration/10 in
+
+      tryDirs $int:spreadProbabilityDenominator;
+      wait spreadableDuration;
+    } {};
+
+    try {
+      create $str:thing;
+      place $str:thing;
+    } {};
+    selfdestruct
+  |]
+  where
+    spreadProbabilityDenominator = floor (1 / spreadProbabilityPerTick) :: Integer
+
+-- | Construct a "combustion robot" from entity, time range and position,
+--   and add it to the world.  It has low priority and will be covered
+--   by placed entities.
+addCombustionBot ::
+  Has (State GameState) sig m =>
+  Combustibility ->
+  Cosmic Location ->
+  TimeSpec ->
+  m ()
+addCombustionBot combustibility loc ts =
+  void $
+    addTRobot $
+      mkRobot
+        ()
+        Nothing
+        "fire"
+        (Markdown.fromText $ T.unwords ["A burning", combustionOutput <> "."])
+        (Just loc)
+        north
+        ( defaultEntityDisplay '*'
+            & displayAttr .~ AWorld "fire"
+            & displayPriority .~ 0
+        )
+        (initMachine combustionProg empty emptyStore)
+        []
+        []
+        True
+        False
+        ts
+  where
+    combustionProg = combustionProgram minT (maxT - minT) combustionOutput spreadLikelihood
+    -- FIXME should support Maybe
+    Combustibility spreadLikelihood (minT, maxT) maybeCombustionProduct = combustibility
+    combustionOutput = fromMaybe "ash" maybeCombustionProduct
```

Comment:
The special default entity name `"ash"` is repeated here and in `defaultCombustibility`.  Let's DRY it up and just have it in one place.